test/stream-for-each.cc

```diff
@@ -0,0 +1,271 @@
+#include "stout/stream-for-each.h"
+
+#include <atomic>
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "stout/collect.h"
+#include "stout/event-loop.h"
+#include "stout/iterate.h"
+#include "stout/let.h"
+#include "stout/map.h"
+#include "stout/range.h"
+#include "stout/stream.h"
+#include "stout/terminal.h"
+#include "stout/then.h"
+#include "stout/timer.h"
+#include "test/event-loop-test.h"
+
+using stout::eventuals::Collect;
+using stout::eventuals::EventLoop;
+using stout::eventuals::Interrupt;
+using stout::eventuals::Iterate;
+using stout::eventuals::Let;
+using stout::eventuals::Map;
+using stout::eventuals::Range;
+using stout::eventuals::Stream;
+using stout::eventuals::StreamForEach;
+using stout::eventuals::Then;
+using stout::eventuals::Timer;
+using testing::ElementsAre;
+
+class StreamForEachTest : public EventLoopTest {};
+
+TEST_F(StreamForEachTest, Interrupt) {
+  auto e = []() {
+    return Iterate(std::vector<int>(1000))
+        | Map(Then([](int x) { return Timer(std::chrono::milliseconds(100)) | Then([x]() { return x; }); }))
+        | StreamForEach([](int x) { return Iterate(std::vector<int>({1, 2})); })
+        | Collect<std::vector<int>>()
+              .stop([](auto& data, auto& k) {
+                k.Start(std::move(data));
+              });
+  };
+
+  auto [future, k] = Terminate(e());
+
+  Interrupt interrupt;
+
+  k.Register(interrupt);
+
+  k.Start();
+
+  interrupt.Trigger();
+
+  EventLoop::Default().Run();
+
+  auto result = future.get();
+
+  CHECK_EQ(result.size(), 0);
+}
+
+TEST(StreamForEach, InterruptReturn) {
+  std::atomic<bool> waiting = false;
+  std::atomic<bool> interrupted = false;
+
+  auto e = [&]() {
+    return Iterate(std::vector<int>(1000))
+        | StreamForEach([&](int x) {
+             return Stream<int>()
+                 .start([&](auto& k) {
+                   waiting.store(true);
+                 })
+                 .next([](auto& k) {
+                   k.Ended();
+                 })
+                 .interrupt([&](auto& k) {
+                   interrupted.store(true);
+                   k.Stop();
+                 });
+           })
+        | Collect<std::vector<int>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  Interrupt interrupt;
+
+  k.Register(interrupt);
+
+  ASSERT_FALSE(waiting.load());
+  ASSERT_FALSE(interrupted.load());
+
+  k.Start();
+
+  ASSERT_TRUE(waiting.load());
+  ASSERT_FALSE(interrupted.load());
+
+  interrupt.Trigger();
+
+  EXPECT_THROW(future.get(), stout::eventuals::StoppedException);
+
+  ASSERT_TRUE(interrupted.load());
+}
+
+// Iterate with init_list
```

Comment:
Is this a TODO that you wanted to change this test to use a `std::initializer_list` instead of a `std::vector`?